File: modules/core/src/main/java/com/panayotis/jubler/subs/style/SubStyle.java
/*
 * (c) 2005-2023 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

package  com.panayotis.jubler.subs.style;

import com.panayotis.jubler.os.DEBUG;
import java.awt.Font;
import java.awt.GraphicsEnvironment;
import java.util.TreeSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SubStyle implements Comparable {

    public static enum Direction {

        TOP, TOPRIGHT, RIGHT, BOTTOMRIGHT, BOTTOM, BOTTOMLEFT, LEFT, TOPLEFT, CENTER
    };
    /* since this is stored already in an array, we will not use StyleType.init() */
    public static final Integer[] FontSizes = {8, 9, 10, 11, 12, 13, 14, 16, 18, 20, 22, 24, 26, 28, 32, 36, 40, 48, 56, 64, 72};
    public static final String[] FontNames;
    private static final Pattern loadpattern;

    static {
        loadpattern = Pattern.compile(
                "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                + "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                + "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                + "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                + "(.*)");

        /* Load font names in a safe manner */
        GraphicsEnvironment env = GraphicsEnvironment.getLocalGraphicsEnvironment();
        String[] fnames;
        try {
            fnames = env.getAvailableFontFamilyNames();
        } catch (Exception e1) {
            DEBUG.debug("Using failsafe routine for font loading.");
            Font[] fnt = env.getAllFonts();
            TreeSet<String> names = new TreeSet<String>();

            for (int i = 0; i < fnt.length; i++)
                try {
                    names.add(fnt[i].getFamily());
                } catch (Exception e2) {
                }

            String[] model = new String[1];
            fnames = names.toArray(model);
        }
        FontNames = fnames;

    }
    public String Name;
    private Object[] values;
    private boolean isDefault = false;

    public SubStyle(SubStyle old) {
        setValues(old);
    }

    /**
     * Creates a new instance of SubStyle
     */
    public SubStyle(String name) {
        this.Name = name;

        /* Initialize default values */
        StyleType[] types = StyleType.values();
        values = new Object[types.length];

        for (int i = 0; i < types.length; i++)
            values[i] = types[i].getDefault();
    }

    public void setName(String newname, SubStyleList list) {
        UniqName uniq = new UniqName(newname);
        Name = uniq.getUniqName(list, this);
    }

    public String getName() {
        return Name;
    }

    public void setValues(SubStyle old) {
        Name = old.Name;
        values = new Object[StyleType.values().length];
        System.arraycopy(old.values, 0, values, 0, values.length);
    }

    public void setValues(String newvalues) {
        if (newvalues == null)
            return;
        Matcher m = loadpattern.matcher(newvalues);
        if (!m.find())
            return;

        String current;
        for (int i = 0; i < values.length - 1; i++) {  // Ignore last "unknown" event
            current = m.group(i + 1);
            values[i] = StyleType.values()[i].init(current);
        }
    }

    public String getValues() {
        StringBuilder out = new StringBuilder();

        for (int i = 0; i < values.length - 1; i++)  // Ignore last "unknown" event
            out.append('|').append(values[i]);
        return out.substring(1);
    }

    public Object get(StyleType which) {
        return values[which.ordinal()];
    }

    public Object get(int which) {
        return values[which];
    }

    public void set(StyleType which, Object what) {
        if (what == null) {
            DEBUG.debug("Ignoring null value found while setting Style " + which.name());
            return;
        }
        int where = which.ordinal();
        /* Overloading doesn't really work well in this case, so we have to force it */
        if (what instanceof String)
            values[where] = which.init((String) what);
        else
            values[where] = which.init(what);
    }

    @Override
    public String toString() {
        return getName();
    }

    public void setDefault(boolean def) {
        isDefault = def;
    }

    public boolean isDefault() {
        return isDefault;
    }

    class UniqName {

        private String text_name = "";
        private int numb_name = 1;
        private String newname = "";

        public UniqName(String name) {
            newname = name;
            int split = name.length() - 1;
            /* Go on and on, until we find a non-numeric character */
            while (split >= 0 && Character.isDigit(name.charAt(split)))
                split--;
            /* Get textual part of the name */
            text_name = name.substring(0, split + 1);

            /* Get numeric part of the name */
            try {
                numb_name = Integer.parseInt(name.substring(split + 1));
            } catch (NumberFormatException e) {
                numb_name = 1;
            }

            /* keep default value, if no numeric part is found */
            if (numb_name < 0)
                numb_name = 1;
        }

        /* Check if the name already exists */
        private boolean findNameInList(NameList list, Object obj) {
            for (int i = 0; i < list.size(); i++)
                if (list.getElementAt(i) != obj && list.getNameAt(i).equals(newname))
                    return true;
            return false;
        }

        private void normalizeInternalName(NameList list, Object obj) {
            /* go through all objects of the list and find the one with the maximum
             * value AND same base text */
            UniqName other;
            for (int i = 0; i < list.size(); i++) {
                /* Split names of list element in two (again) */
                other = new UniqName(list.getNameAt(i));
                if (list.getElementAt(i) != obj && other.numb_name >= numb_name && other.text_name.equals(text_name))
                    /* update number */
                    numb_name = other.numb_name + 1;
            }
        }

        public String getUniqName(NameList list, Object obj) {
            if (findNameInList(list, obj)) {
                normalizeInternalName(list, obj);
                return text_name + numb_name;
            }
            return newname;
        }
    }

    public int compareTo(Object o) {
        return Name.compareTo(((SubStyle) o).Name);
    }
}

File: src/core/style/SubStyle.cpp
/*
 * (c) 2005-2026 by Panayotis Katsaloulis
 * SPDX-License-Identifier: AGPL-3.0-only
 * This file is part of Jubler.
 */

#include "core/style/SubStyle.h"

#include <QRegularExpression>

#include "core/options/Prefs.h"

const QList<int> SubStyle::FontSizes = {8, 9, 10, 11, 12, 13, 14, 16, 18, 20, 22, 24, 26,
                                        28, 32, 36, 40, 48, 56, 64, 72};

namespace {
QStringList g_fontNames;
}

QStringList SubStyle::fontNames() { return g_fontNames; }
void SubStyle::setFontNames(const QStringList &names) { g_fontNames = names; }

SubStyle::SubStyle(const QString &name) : name_(name) {
    for (int i = 0; i < StyleType::COUNT; ++i)
        values_[i] = StyleType::defaultValue(StyleType::Id(i));
}

SubStyle::SubStyle(const SubStyle &old) {
    setValues(old);
}

void SubStyle::setValues(const SubStyle &old) {
    name_ = old.name_;
    values_ = old.values_;
    formatData_ = old.formatData_;
    // isDefault is deliberately not copied (Java copies only Name + values).
}

void SubStyle::setValues(const QString &packed) {
    if (packed.isNull())
        return;
    // Java: 21 lazy groups separated by '|' — i.e. the text must contain at
    // least 20 separators; extra content goes into the last (ignored) group.
    static const QRegularExpression loadpattern(
        QStringLiteral("(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                       "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                       "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                       "(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|(.*?)\\|"
                       "(.*)"));
    const QRegularExpressionMatch m = loadpattern.match(packed);
    if (!m.hasMatch())
        return;
    for (int i = 0; i < StyleType::COUNT - 1; ++i)  // ignore last "unknown" event
        values_[i] = StyleType::init(StyleType::Id(i), m.captured(i + 1));
}

QString SubStyle::getValues() const {
    QString out;
    for (int i = 0; i < StyleType::COUNT - 1; ++i) {  // ignore last "unknown" event
        if (i) out += QLatin1Char('|');
        out += styleValueToString(values_[i]);
    }
    return out;
}

void SubStyle::set(StyleType::Id which, const StyleValue &what) {
    values_[which] = StyleType::init(which, what);
}

void SubStyle::set(StyleType::Id which, const QString &what) {
    values_[which] = StyleType::init(which, what);
}

// ---- UniqName ---------------------------------------------------------------

namespace {
struct UniqName {
    QString textName;
    int numbName = 1;
    QString newName;

    explicit UniqName(const QString &name) : newName(name) {
        int split = name.length() - 1;
        while (split >= 0 && name.at(split).isDigit())
            --split;
        textName = name.left(split + 1);
        bool ok = false;
        numbName = name.mid(split + 1).toInt(&ok);
        if (!ok || numbName < 0)
            numbName = 1;
    }

    bool findNameInList(const SubStyleList &list, const SubStyle *obj) const {
        for (int i = 0; i < list.size(); ++i)
            if (list.get(i).get() != obj && list.getNameAt(i) == newName)
                return true;
        return false;
    }

    void normalizeInternalName(const SubStyleList &list, const SubStyle *obj) {
        for (int i = 0; i < list.size(); ++i) {
            UniqName other(list.getNameAt(i));
            if (list.get(i).get() != obj && other.numbName >= numbName && other.textName == textName)
                numbName = other.numbName + 1;
        }
    }

    QString getUniqName(const SubStyleList &list, const SubStyle *obj) {
        if (findNameInList(list, obj)) {
            normalizeInternalName(list, obj);
            return textName + QString::number(numbName);
        }
        return newName;
    }
};
}  // namespace

void SubStyle::setName(const QString &newName, const SubStyleList &list) {
    UniqName uniq(newName);
    name_ = uniq.getUniqName(list, this);
}

// ---- SubStyleList -----------------------------------------------------------

namespace {
std::unique_ptr<SubStyle> g_defaultStyle;

SubStyle &defaultStyleMutable() {
    if (!g_defaultStyle) {
        g_defaultStyle = std::make_unique<SubStyle>(QStringLiteral("Default"));
        g_defaultStyle->setDefault(true);
        g_defaultStyle->setValues(Prefs::getString(QStringLiteral("styles.default"), QString()));
    }
    return *g_defaultStyle;
}
}  // namespace

const SubStyle &SubStyleList::defaultStyle() {
    return defaultStyleMutable();
}

void SubStyleList::reloadDefaultStyle() {
    g_defaultStyle.reset();
}

SubStyleList::SubStyleList() {
    list_.append(std::make_shared<SubStyle>(defaultStyle()));
    list_.first()->setDefault(true);
}

SubStyleList::SubStyleList(const SubStyleList &old) {
    *this = old;
}

SubStyleList &SubStyleList::operator=(const SubStyleList &old) {
    if (this == &old)
        return *this;
    list_.clear();
    for (const SubStylePtr &s : old.list_)
        list_.append(std::make_shared<SubStyle>(*s));
    if (!list_.isEmpty())
        list_.first()->setDefault(true);
    return *this;
}

int SubStyleList::indexOfName(const QString &name) const {
    for (int i = 0; i < list_.size(); ++i)
        if (name == list_.at(i)->getName())
            return i;
    return -1;
}

int SubStyleList::findStyleIndex(const QString &name) const {
    const int i = indexOfName(name);
    return i < 0 ? 0 : i;
}

SubStylePtr SubStyleList::clearList() {
    SubStylePtr d = list_.isEmpty() ? SubStylePtr() : list_.first();
    list_.clear();
    return d;
}
